service/storage_service.cc

```diff
@@ -5377,6 +5378,96 @@ future<> storage_service::update_fence_version(token_metadata::version_t new_ver
     });
 }
 
+static
+locator::tablet_replica get_leaving_replica(const locator::tablet_info& tinfo, const locator::tablet_transition_info& trinfo) {
+    std::unordered_set<locator::tablet_replica> leaving(tinfo.replicas.begin(), tinfo.replicas.end());
+    for (auto&& r : trinfo.next) {
+        leaving.erase(r);
+    }
+    if (leaving.empty()) {
+        throw std::runtime_error(format("No leaving replicas"));
+    }
+    if (leaving.size() > 1) {
+        throw std::runtime_error(format("More than one leaving replica"));
+    }
+    return *leaving.begin();
+}
+
+inet_address storage_service::host2ip(locator::host_id host) {
```

Comment:
> Could we avoid extending storage_service header by making this a static function, e.g.
static inet_address host2ip(const raft_address_map& map, locator::host_id host)?

If we fear extending storage_service, we should pimpl-ify it instead. This gives freedom to host2ip implementation in choosing dependencies and leaves usage sites clear of clutter.  

> have you considered reusing this?

I considered it, but it would introduce a preemption point which breaks atomicity of the code block, which is a problem. It also is unnecessary to loop here because read barrier ensures that local node knows about ips of all nodes in the topology due to the thing in topology_state_load(). And if it doesn't, streaming will be retried by the coordinator.